File: web/js/services/simulationEngine.js
import { OrbitalState } from '../models/orbitalState.js';

// Drives the simulation loop using requestAnimationFrame.
export class SimulationEngine {
  static OrbitPeriodDays = 29.5;
  static BaseSecondsPerOrbit = 30.0;

  constructor() {
    this.state = new OrbitalState();
    this._isRunning = false;
    this._animFrameId = null;
    this._lastTimestamp = null;
    this.speed = 1.0;
    this.onTick = null;
  }

  get isRunning() {
    return this._isRunning;
  }

  start() {
    if (this._animFrameId !== null) return;
    this._isRunning = true;
    this._lastTimestamp = null;
    this._animFrameId = requestAnimationFrame((t) => this._loop(t));
  }

  togglePause() {
    if (this._isRunning) {
      this._isRunning = false;
      if (this._animFrameId !== null) {
        cancelAnimationFrame(this._animFrameId);
        this._animFrameId = null;
      }
    } else {
      this._isRunning = true;
      this._lastTimestamp = null;
      this._animFrameId = requestAnimationFrame((t) => this._loop(t));
    }
  }

  _loop(timestamp) {
    if (!this._isRunning) return;

    if (this._lastTimestamp !== null) {
      const dtSeconds = (timestamp - this._lastTimestamp) / 1000.0;
      // Cap dt to avoid huge jumps on tab switch
      const clampedDt = Math.min(dtSeconds, 0.1);
      this._tick(clampedDt);
    }
    this._lastTimestamp = timestamp;
    this._animFrameId = requestAnimationFrame((t) => this._loop(t));
  }

  _tick(dt) {
    const degreesPerSecond = 360.0 / SimulationEngine.BaseSecondsPerOrbit;

    // Advance moon angle
    const deltaDegrees = degreesPerSecond * this.speed * dt;
    this.state.moonAngleDegrees =
      (this.state.moonAngleDegrees + deltaDegrees) % 360.0;

    // Advance Earth rotation — Earth spins ~29.5 times per lunar orbit
    const earthDegreesPerSecond =
      degreesPerSecond * SimulationEngine.OrbitPeriodDays;
    const earthDelta = earthDegreesPerSecond * this.speed * dt;
    this.state.earthRotationDegrees =
      (this.state.earthRotationDegrees + earthDelta) % 360.0;

    // Track elapsed days
    const daysPerSecond =
      SimulationEngine.OrbitPeriodDays / SimulationEngine.BaseSecondsPerOrbit;
    this.state.elapsedDays += daysPerSecond * this.speed * dt;

    if (this.onTick) this.onTick();
  }
}

File: MoonSimulation/Services/SimulationEngine.cs
namespace MoonSimulation.Services;

/// <summary>
/// Drives the simulation loop — advances the moon's orbital position
/// at a configurable speed and notifies subscribers on each tick.
/// </summary>
public class SimulationEngine : ISimulationEngine
{
    private IDispatcherTimer? _timer;
    private readonly Models.OrbitalState _state = new();
    private bool _isRunning;

    // One full orbit = 29.5 Earth days.  At 1x speed we complete it in 120 seconds real-time.
    private const double OrbitPeriodDays = 29.5;
    private const double BaseSecondsPerOrbit = 120.0;
    private const double TickIntervalSeconds = 1.0 / 60.0; // 60 fps

    /// <summary>Speed multiplier: 0.1 (slow) → 3 (fast).</summary>
    public double Speed { get; set; } = 0.5;

    public Models.OrbitalState State => _state;
    public bool IsRunning => _isRunning;

    /// <summary>Fired every tick so views can redraw.</summary>
    public event Action? OnTick;

    public void Start(IDispatcher dispatcher)
    {
        if (_timer != null) return;

        _timer = dispatcher.CreateTimer();
        _timer.Interval = TimeSpan.FromSeconds(TickIntervalSeconds);
        _timer.Tick += (_, _) => Tick();
        _timer.Start();
        _isRunning = true;
    }

    public void TogglePause()
    {
        if (_timer == null) return;
        if (_isRunning)
        {
            _timer.Stop();
            _isRunning = false;
        }
        else
        {
            _timer.Start();
            _isRunning = true;
        }
    }

    private void Tick()
    {
        // Degrees per second at 1x speed
        double degreesPerSecond = 360.0 / BaseSecondsPerOrbit;

        // Advance moon angle
        double deltaDegrees = degreesPerSecond * Speed * TickIntervalSeconds;
        _state.MoonAngleDegrees = (_state.MoonAngleDegrees + deltaDegrees) % 360.0;

        // Advance Earth rotation — Earth spins ~29.5 times per lunar orbit
        double earthDegreesPerSecond = degreesPerSecond * OrbitPeriodDays;
        double earthDelta = earthDegreesPerSecond * Speed * TickIntervalSeconds;
        _state.EarthRotationDegrees = (_state.EarthRotationDegrees + earthDelta) % 360.0;

        // Track elapsed days
        double daysPerSecond = OrbitPeriodDays / BaseSecondsPerOrbit;
        _state.ElapsedDays += daysPerSecond * Speed * TickIntervalSeconds;

        OnTick?.Invoke();
    }
}
